File: kotlin/backtracking/main_14888.kt
import kotlin.math.max
import kotlin.math.min

lateinit var num: IntArray
lateinit var operation: IntArray
var n = 0
var minValue = 10_0000_0000
var maxValue = -10_0000_0000

fun main() {
    n = readLine()!!.toInt()
    num = readLine()!!.split(' ').map { it.toInt() }.toIntArray()
    operation = readLine()!!.split(' ').map { it.toInt() }.toIntArray()
    calc(1, num[0])
    println(maxValue)
    println(minValue)
}

fun calc(depth: Int, value: Int) {
    if (depth == n) {
        minValue = min(minValue, value)
        maxValue = max(maxValue, value)
    } else {
        for (i in 0..3){
            if (operation[i] > 0) {
                operation[i]--
                calc(depth + 1, when(i) {
                    0 -> value + num[depth]
                    1 -> value - num[depth]
                    2 -> value * num[depth]
                    else -> value / num[depth]
                })
                operation[i]++
            }
        }
    }
}

File: cpp/step/backtracking/main_14888.cpp
#include <iostream>
#include <cmath>
using namespace std;

int n;
int arr[11];
int operation[4];
int minValue = 1000000000;
int maxValue = -1000000000;

void dfs(int depth, int calc) {
	if (depth == n) {
		minValue = min(minValue, calc);
		maxValue = max(maxValue, calc);
	} else {
		for (int i = 0; i < 4; ++i) {
			if (operation[i] > 0) {
				operation[i]--;
				int tem = calc;
				if (i == 0) {
					tem += arr[depth];
				} else if (i == 1) {
					tem -= arr[depth];
				} else if (i == 2) {
					tem *= arr[depth];
				} else {
					tem /= arr[depth];
				}
				dfs(depth + 1, tem);
				operation[i]++;
			}
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	cin >> n;
	for (int i = 0; i < n; ++i) {
		cin >> arr[i];
	}
	for (int& op : operation) {
		cin >> op;
	}
	dfs(1, arr[0]);
	cout << maxValue << '\n' << minValue;
}
